src/main/telemetry/msp_shared.c

```diff
@@ -29,225 +29,294 @@
+STATIC_UNIT_TESTED struct mspPacket_s requestPacket;
+STATIC_UNIT_TESTED struct mspPacket_s responsePacket;
+
 static mspDescriptor_t mspSharedDescriptor;
 
 void initSharedMsp(void)
 {
-    mspPackage.requestBuffer = (uint8_t *)&mspRxBuffer;
-    mspPackage.requestPacket = &mspRxPacket;
-    mspPackage.requestPacket->buf.ptr = mspPackage.requestBuffer;
-    mspPackage.requestPacket->buf.end = mspPackage.requestBuffer;
-
-    mspPackage.responseBuffer = (uint8_t *)&mspTxBuffer;
-    mspPackage.responsePacket = &mspTxPacket;
-    mspPackage.responsePacket->buf.ptr = mspPackage.responseBuffer;
-    mspPackage.responsePacket->buf.end = mspPackage.responseBuffer;
+    responsePacket.buf.ptr = mspSerialOutBuf;
+    responsePacket.buf.end = ARRAYEND(mspSerialOutBuf);
 
     mspSharedDescriptor = mspDescriptorAlloc();
 }
 
 static void processMspPacket(void)
 {
-    mspPackage.responsePacket->cmd = 0;
-    mspPackage.responsePacket->result = 0;
-    mspPackage.responsePacket->buf.end = mspPackage.responseBuffer;
+    responsePacket.cmd = 0;
+    responsePacket.result = 0;
+    responsePacket.buf.ptr = mspSerialOutBuf;
+    responsePacket.buf.end = ARRAYEND(mspSerialOutBuf);
 
     mspPostProcessFnPtr mspPostProcessFn = NULL;
-    if (mspFcProcessCommand(mspSharedDescriptor, mspPackage.requestPacket, mspPackage.responsePacket, &mspPostProcessFn) == MSP_RESULT_ERROR) {
-        sbufWriteU8(&mspPackage.responsePacket->buf, TELEMETRY_MSP_ERROR);
+    if (mspFcProcessCommand(mspSharedDescriptor, &requestPacket, &responsePacket, &mspPostProcessFn) == MSP_RESULT_ERROR) {
+        sbufWriteU8(&responsePacket.buf, TELEMETRY_MSP_ERROR);
     }
     if (mspPostProcessFn) {
         mspPostProcessFn(NULL);
     }
 
-    sbufSwitchToReader(&mspPackage.responsePacket->buf, mspPackage.responseBuffer);
+    sbufSwitchToReader(&responsePacket.buf, mspSerialOutBuf);
 }
 
 void sendMspErrorResponse(uint8_t error, int16_t cmd)
 {
-    mspPackage.responsePacket->cmd = cmd;
-    mspPackage.responsePacket->result = 0;
-    mspPackage.responsePacket->buf.end = mspPackage.responseBuffer;
+    responsePacket.cmd = cmd;
+    responsePacket.result = 0;
+    responsePacket.buf.ptr = mspSerialOutBuf;
 
-    sbufWriteU8(&mspPackage.responsePacket->buf, error);
-    mspPackage.responsePacket->result = TELEMETRY_MSP_RES_ERROR;
-    sbufSwitchToReader(&mspPackage.responsePacket->buf, mspPackage.responseBuffer);
+    sbufWriteU8(&responsePacket.buf, error);
+    responsePacket.result = TELEMETRY_MSP_RES_ERROR;
+    sbufSwitchToReader(&responsePacket.buf, mspSerialOutBuf);
 }
 
-bool handleMspFrame(uint8_t *frameStart, int frameLength, uint8_t *skipsBeforeResponse)
+// despite its name, the function actually handles CRSF frame payload with MSP in it
+// it reads the MSP into requestPacket stucture and handles it after receiving all the chunks.
+bool handleMspFrame(uint8_t *crsfPayload, uint8_t crsfPayloadLength, uint8_t *skipsBeforeResponse)
```

Comment:
Should just be ```payload``` and ```payloadLength```. As mentioned earlier, this isn't specific to crossfire. It's used for sbus/smartport/fport too.